File: backend/middleware/roles.js
// middleware/roles.js
// ─────────────────────────────────────────────────────────
// يتحقق من صلاحيات المستخدم — يستخدم بعد verifyToken
// الاستخدام: router.get('/...', verifyToken, requireRole('admin'), handler)
// ─────────────────────────────────────────────────────────

function requireRole(...roles) {
  return (req, res, next) => {
    // req.user أتى من middleware/auth.js
    if (!req.user || !roles.includes(req.user.role)) {
      return res.status(403).json({
        error: `هذه الصفحة للـ ${roles.join(' / ')} فقط — ليس لديك صلاحية`
      });
    }
    next();
  };
}

module.exports = requireRole;

File: backend-py/middleware/roles.py
from fastapi import HTTPException


def require_role(user, *roles):
    if not user or user.get("role") not in roles:
        raise HTTPException(
            status_code=403,
            detail=f"هذه الصفحة للـ {' / '.join(roles)} فقط — ليس لديك صلاحية",
        )


def require_permission(user, section):
    """تحقق من صلاحية المستخدم للوصول إلى قسم معين.
    admin: مسموح دائماً.
    permissions = NULL: مسموح (غير محدود — توافق مع الحسابات القديمة).
    permissions = قائمة: يجب أن تحتوي على section.
    """
    if not user:
        raise HTTPException(status_code=403, detail="غير مسموح")
    if user.get("role") == "admin":
        return
    permissions = user.get("permissions")
    if permissions is None:
        return
    if section not in permissions:
        raise HTTPException(status_code=403, detail="ليس لديك صلاحية الوصول لهذا القسم")
